blogposts/2021/how-to-not-break-a-search-engine-or-what-i-learned-about-unglamorous-engineering.md

```diff
@@ -0,0 +1,290 @@
+queries stopped working? What if they behaved incorrectly? What if an input crashed
+the server? The point was to support new operators in the query, and I was
+going to introduce these, but mentally I was more preoccupied with ensuring
+that _existing queries keep working_ the way that they should.
+
+Writing a parser isn't _that_ difficult. But it's a different matter to write
+one from scratch for a fresh project versus swapping out something that is in
+use by thousands of users. When the stakes are high, simple things become
+challenging. This code will need checking and checking again. Not exactly
+glamorous.
+
+The changes weren't just about the act of parsing either. The new parser would
+produce a different internal representation of a query. Trees would represent
+our queries now, no longer just a sequential list of terms. This new
+representation needed to keep working with what our backend internals expect.
+That, or bugs galore.
+
+I made up my mind that, as far as I could help it, swapping our parsers wasn't
+going to break Sourcegraph. Not on my account.
+
+## How not to break a search engine
+
+Over the course of development I progressively tested new changes in four
+distinct ways. I wasn't following a checklist of things to do. What mattered
+was that I could convince myself that things worked as expected. I wanted
+to sleep well at night. Testing along these four different dimensions gave me
+confidence that I could expect few (if any) surprises. In general, the sort of
+testing for application rewrites or migration is going to depend on the
+context. I feel like the steps I took worked well, and they probably generalize
+well for migrating other systems that accept user input.
+
+### Part 1: Unit testing
+
+Unit testing was the obvious thing to do™ and familiar territory for a lot of
+developers, so I won't belabor this part. I thought of this as my frontline
+defense for testing correctness. I reused some of our existing parser tests
+(which served as a rough specification of things to get right) and also added a
+lot of additional tests for new parts of the syntax. You bet there's test
+coverage.
+
+
+<img src="https://storage.googleapis.com/sourcegraph-assets/about.sourcegraph.com/blog/2021/parser-coverage.png" style="width: 300px">
+
+### Part 2: Integration testing
+
+Once I wrote a feature-flagged code path to use the new parser, it was time to
+test whether Sourcegraph behaved functionally similar. Integration tests ran a
+search query through a running instance of the Sourcegraph application and
+checked whether search results were expected. This testing phase was so
+important because the parser produced a new and different internal
+representation. I'd abstracted out a common interface for our backend to access
+the new data structure under the feature-flagged code path to test on.
+
+<img src="https://storage.googleapis.com/sourcegraph-assets/about.sourcegraph.com/blog/2021/integration.png" style="width: 800px">
```

Comment:
@rvantonder Forgot to mention – I don't think every image needs a caption, but for ones like this that show a bit more detail, I think it's helpful to add one briefly describing what the image is showing. Some people will skim the blog post and being able to see easily what the image is about can help to draw them in. You can see some examples here: https://about.sourcegraph.com/blog/optimizing-a-code-intel-commit-graph/

Reply:
done. also added `alt` to all imgs